### ISSUE TITLE
Port xsd.exe to .NET Core

### ISSUE THREAD

**Issue by hrushib** [dotnet/runtime#1391]
We use xsd.exe to generate C# classes for based on xml schema. At present there is no xsd.exe tool to generate CoreCLR compliant code and hence code generation can not be fully automated as part of the build at the moment. 

Please let us know plan/timelines around availability of the same.

**Comment by StephenBonikowsky** [comment]
Increasing the priority for this task.

**Comment by theRainbird** [comment]
I need to create typed DataSet code, because I'm planning to port a very large ERP application (over 1 million lines of code) to .NET Core, which massively uses typed DataSets. Without xsd.exe or another way to create type DataSet code, a port is not possible.

**Comment by mv10-work** [comment]
@theRainbird With all due respect, I hope they do not implement `DataSet` support, it's a very outdated model. I just finished migrating a huge, messy legacy application off `DataSet` to more sensible generics and LINQ, so I sympathize with you, but from what I've seen, that pain-point hasn't been a convincing argument for adding old-world features to modern .NET. (If they do implement it, I hope they finally fix the atrocious [`DBNull` exception mess](https://social.msdn.microsoft.com/Forums/en-US/2e1c4146-994a-4ebb-b090-fd271aebb880/you-can-have-stronglytyped-datatables-or-you-can-have-nullable-columns-choose-one?forum=adodotnetdataset) introduced way back in the .NET 2.0 release.)

**Comment by wilfriedb** [comment]
TL;DR: I'm trying to get xsd.exe to run under .NET 6.0. As a prerequisite, I've made some changes to XML.Serialization. Am I heading in the right way?

Because our team also has a need for for xsd.exe (or a similar tool) I went looking what is needed to revive xsd.exe. I found the source code on [ReferenceSource](https://github.com/microsoft/referencesource), and tried to build it against .NET Core 5.0. This didn't work because one type was missing: XmlCodeExporter. So I forked dotnet/runtime and copied the missing type from .NET 4.8 into it. Of course, this caused a small cascade of other missing types, and I added them as well. After adding a lot of nullable annotations and null checks I now have a version of XML.Serialization that builds with these old/new types included.

My work so far:
[XmlCodeExporter add](https://github.com/wilfriedb/runtime/tree/add-code-exporter-to-xml-serializer)

There's still some work to be done: I couldn't find the resource strings in ReferenceSource (maybe someone else can find them), unit tests need to be added, and the coding style of the copied code is really old fashioned, maybe that could be cleaned up as well.

Before I spend more time in this effort, Is this modification (actually, just re-adding some classes) something that the .NET team will be going to accept a pull request for? Or has the team another solution in mind?

I would love to hear from you.

**Comment by danmoseley** [comment]
@HongGit

**Comment by theRainbird** [comment]
> @theRainbird With all due respect, I hope they do not implement `DataSet` support, it's a very outdated model.

Nobody forces you to use this old school stuff.
But if you have to migrate a large application with over a million lines of code to .NET 5+, it is very bad if you have to write everything from scratch.

**Comment by davidxavier77** [comment]
Hi,
  dotnet core 6.0  is out. Is it in or out  ?

**Comment by StephenMolloy** [comment]
As @wilfriedb points out [above](https://github.com/dotnet/runtime/issues/1391#issuecomment-808747826), the source code for xsd.exe in the .Net Framework is open sourced. If anyone wants to take a stab at a .Net Core implementation, our team is unable to prioritize this at this time.

**Comment by danmoseley** [comment]
> I couldn't find the resource strings in ReferenceSource (maybe someone else can find them), 

https://github.com/microsoft/referencesource/blob/master/xsd/XsdRes.txt

> coding style of the copied code is really old fashioned,

As you can probably imagine, that's quite typical of the C# written around .NET 1.0, 20 years ago now. 🙂 The style you see in .NET Core repos was cleaned up substantially when the project began.

**Comment by sanghel-payoff** [comment]
... maybe the kind soul that tackles this will event fix https://github.com/dotnet/wcf/issues/1274

**Comment by jinek** [comment]
Second option is to use [mono xsd tool](https://manpages.debian.org/stretch/mono-devel/xsd.1)

**Comment by deeprobin** [comment]
Is it at all still relevant to have a tool like `xsd.exe` which exports the code to files?

That would be a good use case for a source generator imo (indicated by a `XsdGeneratorAttribute`),

**Comment by Lancelotbronner** [comment]
I still like the idea of a CLI tool to work with xsd, I'll try and modernize it but if it needs to touch the runtime then I don't think I'll be able to do much.

**Comment by LevYas** [comment]
@deeprobin 
> Is it at all still relevant to have a tool like `xsd.exe` which exports the code to files?
Sure, almost all the XSD I had were not 100% correct, I had to modify the resulting classes to make them actually work with the third-party APIs.
From a tool like xsd.exe, I expect to produce compact and readable code, which I can reason about and modify if needed.

**Comment by soroshsabz** [comment]
ITNOA

Any update? does XSD.exe source code available?

**Comment by soroshsabz** [comment]
Ok I found it https://github.com/microsoft/referencesource/tree/master/xsd

:D

**Comment by nalinasanka** [comment]
Hi All,

I was able to generate XSD file for .Net Core library using below code. If anyone is keen to generate xsd file for .Net Core instead of using "xsd.exe", hope this code will be handy.

```
try
{
    XmlSchemas schemas = new XmlSchemas();
    XmlSchemaExporter exporter = new XmlSchemaExporter(schemas);

    XmlTypeMapping mapping = new XmlReflectionImporter().ImportTypeMapping(typeof("Your Class Name"));
    exporter.ExportTypeMapping(mapping);

    using (TextWriter writer = new StreamWriter("output.xsd"))
    {
        foreach (XmlSchema schema in schemas)
        {
            schema.Write(writer);
        }
    }

    Console.WriteLine("XSD generated successfully.");
}
catch (Exception ex)
{
    Console.WriteLine($"Error in generating XSD. Error: {ex.Message}");
}
```

"Your Class Name" --> Needs to replace with your class name which you want to generate XSD schema.

Thanks

**Comment by deeprobin** [comment]
@StephenMolloy @HongGit Maybe it fits in this month that I start an attempt for an xsd prototype.
Could you assign me to the feature?

Should we create a new repo for it as a separate tool or what would you suggest?

**Comment by LevYas** [comment]
Fun fact, since my recent comment I changed the way I generate DTOs. I feed an example of XML and XSD schema into ChatGpt and ask it to write minimalistic DTO for it, and it's very easy to control what attributes are needed and what formatting style should be, what classes should be first. That works surprisingly well, so I abandoned all conversion utilities.

**Comment by leus** [comment]
Hello,

We still need to consume / produce files based on XSD schemas that the current version of xsd.exe cannot handle (xsd:import, xsd:include, etc.). Is anybody working this? I know from bad experiences in the past that contributing code to dotnet is a minefield, so I wonder if it is worth to give a try (I see a`help wanted` tag).
